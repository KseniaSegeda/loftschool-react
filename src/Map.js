import React from "react";

const Map = () => (
    <h2>Map</h2>
)

export default Map;
