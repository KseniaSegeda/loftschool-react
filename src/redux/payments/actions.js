import { createAction } from 'redux-actions';

export const pendingSetCard = createAction('PENDING_SET_CARD');
export const successSetCard = createAction('SUCCESS_SET_CARD');


export const pendingGetCard = createAction('PENDING_GET_CARD');
export const successGetCard = createAction('SUCCESS_GET_CARD');

export const errorCard = createAction('ERROR_CARD');
