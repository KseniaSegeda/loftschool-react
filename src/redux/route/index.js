export * from "./actions";
export * from "./saga";
export * from "./select";
export { default } from './reducer';
