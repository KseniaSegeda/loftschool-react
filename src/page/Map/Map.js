import React, {Component} from "react";
import mapboxgl from "mapbox-gl";
import "./Map.css";
mapboxgl.accessToken = 'pk.eyJ1IjoiZHJhZ2luIiwiYSI6ImNraXpob3drMDFvNmcyem56NDI1c2prMTIifQ.zfgJa232mWsv5DD2AQmd_w';


class Map extends Component {
    constructor(props) {
        super(props);
        this.map = null;
        this.mapContainer = React.createRef();
    }

    componentDidMount() {
        this.map = new mapboxgl.Map({
            container: this.mapContainer.current,
            style: "mapbox://styles/mapbox/streets-v11",
            center: [37.61513, 55.7513461], // moscow
            zoom: 11,
        });
    }

    render(){
        return <div data-testid="map" className="map-container" ref={this.mapContainer} />;
    }
}


export default Map;
