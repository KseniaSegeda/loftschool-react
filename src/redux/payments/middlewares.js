import {pendingSetCard, successSetCard, errorCard} from "./actions";
import {serverSetCard} from "../../api/card";

export const payMiddleware = (store) => (next) => async (action) => {
    switch (action.type) {
        case pendingSetCard.toString():
            const token = JSON.parse(window.localStorage.getItem('state')).token;
            const responseSetCard = await serverSetCard({...action.payload, token});

            if (responseSetCard.success) {
                store.dispatch(successSetCard(action.payload))
            } else {
                store.dispatch(errorCard(responseSetCard.error))
            }
            break;
        default:
            next(action);
    }
}
