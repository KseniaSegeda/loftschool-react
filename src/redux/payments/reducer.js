
import {successSetCard, errorCard, pendingGetCard} from "./actions";
const initialState = {
    cardNumber: '',
    expiryDate: '',
    cardName: '',
    cvc:'',
    isLoading: false
}

// eslint-disable-next-line import/no-anonymous-default-export
export default function(state = initialState, action) {
    switch(action.type){
        case pendingGetCard.toString():
            return {
                isLoading: false
            }
        case successSetCard.toString():
            const {cardNumber, expiryDate, cardName, cvc} = action.payload;
            return {
                cardNumber,
                expiryDate,
                cardName,
                cvc,
                isLoading: true
            }
        case errorCard.toString():
            return {
                isLoading: true,
                error: action.payload
            }
        default:
            return state;
    }
}

export const getCard = (state) => {
    const {
        cardNumber,
        expiryDate,
        cardName,
        cvc,
        isLoading
    } = state.pay;
    return {
        cardNumber,
        expiryDate,
        cardName,
        cvc,
        isLoading
    };
}
