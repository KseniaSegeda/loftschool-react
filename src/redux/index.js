import { combineReducers } from 'redux';
import auth from './auth/reducer';
import pay from './payments/reducer';
import route from './route/reducer'

export default combineReducers({auth, pay, route})
