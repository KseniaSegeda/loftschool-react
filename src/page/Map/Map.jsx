import React, {useEffect, useRef, useState} from "react";
import OpenOrder from "../../components/OpenOrder/OpenOrder"
import mapboxgl from "mapbox-gl";
import "./Map.css";
import {getRoute} from "../../redux/route"
import {connect} from "react-redux";
import PaymentsOrder from "../../components/PaymentsOrder/PaymentsOrder";

const Map = (props) => {
    const {addresses, draw, isCard} = props;

    const [modal, setModal] = useState('payments');

    mapboxgl.accessToken = 'pk.eyJ1IjoiZHJhZ2luIiwiYSI6ImN' +
        'raXpob3drMDFvNmcyem56NDI1c2prMTIifQ.zfgJa232mWsv5DD2AQmd_w';
    const mapContainer = useRef(null);
    useEffect(() => {
        const myMap = new mapboxgl.Map({
            container: mapContainer.current,
            style: "mapbox://styles/mapbox/streets-v11",
            center: [30.302499027775248, 59.92261297636499],
            zoom: 11,
        });
        !isCard? setModal('payments') : setModal('order');

        if(draw.drawRoute.length && !draw.isLoading){
            const coordinates = draw.drawRoute;
            myMap.on('styledata', function() {drawRoute(myMap, coordinates)});
        }
        // return () => myMap.remove();
    }, [draw.drawRoute, draw.isLoading, isCard]);
    return <>
        {modal=== 'payments' ?  <PaymentsOrder/> : null}
        {modal=== 'order' ? <OpenOrder listAddress={addresses} />: null}
        <div data-testid="map" className="map-container" ref={mapContainer} />
    </>;
}

export const drawRoute = (map, coordinates) => {
    map.flyTo({
        center: coordinates[0],
        zoom: 15
    });

    map.addLayer({
        id: "route",
        type: "line",
        source: {
            type: "geojson",
            data: {
                type: "Feature",
                properties: {},
                geometry: {
                    type: "LineString",
                    coordinates
                }
            }
        },
        layout: {
            "line-join": "round",
            "line-cap": "round"
        },
        paint: {
            "line-color": "blue",
            "line-width": 8
        }
    });
};

export default connect(
    getRoute,
    {}
)(Map);
