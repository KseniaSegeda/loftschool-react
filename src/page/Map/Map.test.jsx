import ReactDOM from "react-dom";
import Map from "../../page/Map/Map";
import React from "react";
import renderer from "react-test-renderer";
import {render, act} from '@testing-library/react'
import mapboxgl from "mapbox-gl";
import {createStore} from "redux";
import rootReducer from "../../redux";
import {successSetCard} from "../../redux/payments";
import {successGetListAddress} from "../../redux/route";

jest.mock('mapbox-gl', () => ({
    GeolocateControl: jest.fn(),
    Map: jest.fn(() => ({
        addControl: jest.fn(),
        on: jest.fn(),
        remove: jest.fn()
    })),
    NavigationControl: jest.fn()
}));
const FAKE_DATA_CARD = {
    cardNumber: '1111 1111 1111 1111',
    expiryDate: '10/24',
    cardName: 'TEST TEST',
    cvc: '111'
}

describe('Map.js', () => {
    let store;
    beforeAll(() => {
        store = createStore(rootReducer);
    })

    it("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(global.wrapperProvider(<Map/>, store), div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('should call mapbox-gl', function () {
        const { getByTestId } = render(global.wrapperProvider(<Map />, store));
        expect(mapboxgl.Map).toHaveBeenCalledWith({
            container: getByTestId('map'),
            style: "mapbox://styles/mapbox/streets-v11",
            center: [30.302499027775248, 59.92261297636499],
            zoom: 11,
        });
    });

    it('render correctly', function () {
        const tree = renderer.create(global.wrapperProvider(<Map />, store)).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('card not set -> popup redirect profile', async function () {
        let  getByText;
        await act(async () => {
            getByText = render(global.wrapperProvider(<Map />, store)).getByText;
        });
        expect(getByText('Заполните платежные данные для заказа')).toBeInTheDocument();
    });

    it('card set -> popup open order', async function () {
        store.dispatch(successSetCard(FAKE_DATA_CARD));
        let  getByText;
        await act(async () => {
            getByText = render(global.wrapperProvider(<Map />, store)).getByText;
        });
        store.dispatch(successGetListAddress({addresses: [1,2,3,4]}));

        expect(getByText('Вызвать такси')).toBeInTheDocument();
    });

    it('card set -> popup open order(loading)', async function () {
        let  getByText;
        store.dispatch(successSetCard(FAKE_DATA_CARD));
        await act(async () => {
            getByText = render(global.wrapperProvider(<Map />, store)).getByText;
        });

        expect(getByText('Loading')).toBeInTheDocument();
    });
})
