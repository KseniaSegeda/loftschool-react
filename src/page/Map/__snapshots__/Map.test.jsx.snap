// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map.js render correctly 1`] = `
Array [
  <div
    className="MuiPaper-root openOrder MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="containerOpenOrder"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        Заполните платежные данные для заказа
      </p>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <a
          href="/profile"
          onClick={[Function]}
        >
          Заполнить
        </a>
      </span>
    </button>
  </div>,
  <div
    className="map-container"
    data-testid="map"
  />,
]
`;
